bindings/rascal/utils/filter_utils.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+
+# utility functions for Filters
+
+
+def get_index_mappings_sample_per_species(managers, sps):
+    # get various info from the structures about the center atom species and indexing
+    types = []
+    strides_by_sp = {sp: [0] for sp in sps}
+    global_counter = {sp: 0 for sp in sps}
+    indices_by_sp = {sp: [] for sp in sps}
+    map_by_manager = {sp:[{} for ii in range(len(managers)) ] for sp in sps}
+    for i_man in range(len(managers)):
+        man = managers[i_man]
+        counter = {sp: 0 for sp in sps}
+        for i_at, at in enumerate(man):
+            types.append(at.atom_type)
+            if at.atom_type in sps:
+                map_by_manager[sp][i_man][global_counter[at.atom_type]] = i_at
+                counter[at.atom_type] += 1
+                global_counter[at.atom_type] += 1
+            else:
+                raise ValueError(
+                    'Atom type {} has not been specified in fselect: {}'.format(
+                    at.atom_type, sps))
+        for sp in sps:
+            strides_by_sp[sp].append(counter[sp])
+
+    for sp in sps:
+        strides_by_sp[sp] = np.cumsum(strides_by_sp[sp])
+
+    for ii, sp in enumerate(types):
+        indices_by_sp[sp].append(ii)
+
+    return strides_by_sp, global_counter, map_by_manager, indices_by_sp
+
+
+def get_index_mappings_sample(managers):
+    # get various info from the structures about the center atom species and indexing
+    strides = [0]
+    global_counter = 0
+    map_by_manager = [{} for ii in range(len(managers))]
+    for i_man in range(len(managers)):
+        man = managers[i_man]
+        counter = 0
+        for i_at, _ in enumerate(man):
+            map_by_manager[i_man][global_counter] = i_at
+            counter += 1
+            global_counter += 1
+        strides.append(counter)
+
+    strides = np.cumsum(strides)
+
+    return strides, global_counter, map_by_manager
+
+
+def convert_selected_global_index2rascal_sample_per_species(managers, selected_ids_by_sp,
```

Comment:
> is this just a global-to-rascal index converter?

Isn't it just what the function name means ?

Reply:
No, it's too long and hard to parse out the actual meaning. Something like `index_convert_global_rascal` might be easier to understand. And could the `per_species` bit be taken care of by an optional argument?

Reply:
> And could the per_species bit be taken care of by an optional argument?

its not very practical I think because the `per_species` calls `convert_selected_global_index2rascal_sample`

Reply:
Ok, then how about this function gets the shorter name (because it's the one that gets called by other modules) and the other one gets called `<convert_index_whatever>_singlespecies`?

Also: What is the "rascal" indexing format? Is it just a per-molecule index? If so, `permolecule` might be a more descriptive name than `rascal`.

Reply:
yes its a `perstructure` indexing.

Reply:
> `_singlespecies` 

it is more like all species are in the same bin while the other routine puts them in different bins.

Reply:
This is getting close to the previous discussion